### PR TITLE
[WIP] Add basic plugin support

### LINKED ISSUE #1048: Custom Middlewares?
Apologize if this is the wrong forum, Is there support for or would be a good feature to support custom injectable middlewares. Traefik looks amazing, we are using Fabio from ebay now and are looking at mantl and found Traefik.

### LINKED ISSUE #1336: enable custom plugins/middlewares for Traefik
After seeing the [Go1.8 new plugin feature](https://golang.org/pkg/plugin/) I though that this could help a lot o people to add specific functionalities to Traefik.

Instead of building/compiling/shipping a custom-made version of Traefik to enable a custom functionality it would be possible to write way simpler custom-made middlewares with this approach, doesn't it ? 

Try imagine creating a package that receives the request at a parameter without having to recompile the whole Traefik repository just to add a small change. Does it sounds like a middleware ? Because for me it is ! It's just a go1.8-plugin-based-middleware !

What do you guys think ?

### PR DESCRIPTION
**Pull request by emilevauge**
This PR is a proposal for adding basic plugin support in Traefik using the new Golang 1.8 plugin mechanism.

Right now, only `Middleware` plugin type is available.
Build your plugin using:

```sh
go build -buildmode=plugin -o sample.so sample.go
```

Then add in you Traefik toml file:

```toml
[[plugins]]
path = "sample.so"
```

Please, give your feedback on this basic implementation :)

/cc @containous/traefik 

Fixes #1336
Fixes #1048
Part of #30

Signed-off-by: Emile Vauge <emile@vauge.com>

### CODE REVIEW

File: plugin/plugin.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+// Plugins defines a set of Plugin
+type Plugins []*Plugin
+
+//Set []*Plugin
+func (p *Plugins) Set(str string) error {
+	exps := strings.Split(str, ",")
+	if len(exps) == 0 {
+		return fmt.Errorf("Bad Plugin format: %s", str)
```
Comment (timoreimann) [inline plugin/plugin.go]: Bad -> bad

File: plugin/manager.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
+package plugin
+
+import (
+	"fmt"
+	goplugin "plugin"
+)
+
+// Manager is in charge of instanciating plugins and storing those in memory
+type Manager struct {
+	middlewares []*Middleware
```
Comment (timoreimann) [inline plugin/manager.go]: Does this have to be a slice of pointers, or could it be simplified to a slice of values?

File: plugin/plugin.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+	"strings"
+)
+
+// Plugin defines a plugin configuration in traefik
+type Plugin struct {
+	Path string
+}
+
+// Plugins defines a set of Plugin
+type Plugins []*Plugin
```
Comment (timoreimann) [inline plugin/plugin.go]: Slice of values possible?

File: plugin/plugin.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+//Set []*Plugin
```
Comment (timoreimann) [inline plugin/plugin.go]: GoDoc should be expanded here and below.

File: plugin/plugin.go [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+func (p *Plugins) String() string { return fmt.Sprintf("%+v", *p) }
+
+//SetValue sets []*Plugin into the parser
+func (p *Plugins) SetValue(val interface{}) {
+	*p = Plugins(val.(Plugins))
+}
+
+// Type exports the Plugins type as a string
+func (p *Plugins) Type() string {
+	return fmt.Sprint("plugins")
```
Comment (timoreimann) [inline plugin/plugin.go]: No need to use `fmt.Sprint`, can return a string value directly.

File: configuration.go [hunk 6]
```diff
@@ -38,6 +39,7 @@ type GlobalConfiguration struct {
 	MaxIdleConnsPerHost       int                     `description:"If non-zero, controls the maximum idle (keep-alive) to keep per-host.  If zero, DefaultMaxIdleConnsPerHost is used"`
 	InsecureSkipVerify        bool                    `description:"Disable SSL certificate verification"`
 	Retry                     *Retry                  `description:"Enable retry sending request if network error"`
+	Plugins                   plugin.Plugins          `description:"External plugins definition"`
```
Comment (timoreimann) [inline configuration.go]: I'd amend that it currently supports middlewares only.

File: plugin/manager.go [hunk 7]
```diff
@@ -0,0 +1,44 @@
+package plugin
+
+import (
+	"fmt"
+	goplugin "plugin"
+)
+
+// Manager is in charge of instanciating plugins and storing those in memory
```
Comment (emilevauge) [inline plugin/manager.go]: fixed

File: plugin/manager.go (hunk 2 above)
Comment (emilevauge) [inline plugin/manager.go]: The whole question is: do we allow stateful plugins?

File: plugin/plugin.go (hunk 3 above)
Comment (emilevauge) [inline plugin/plugin.go]: yep, done ;)

File: plugin/plugin.go (hunk 1 above)
Comment (emilevauge) [inline plugin/plugin.go]: Done

File: plugin/plugin.go (hunk 5 above)
Comment (emilevauge) [inline plugin/plugin.go]: Hum, don't know why this happened ^^

File: configuration.go (hunk 6 above)
Comment (emilevauge) [inline configuration.go]: I will give more details in the documentation. I prefer to keep flags description simple :)

File: plugin/manager.go [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+	if err != nil {
+		return fmt.Errorf("Error in plugin Lookup: %s", err)
+	}
+	instance := loader.(func() interface{})()
+	middleware, ok := instance.(Middleware)
+	if ok {
+		// if Middleware, add to middleware list
+		m.middlewares = append(m.middlewares, &middleware)
+	} else {
+		return fmt.Errorf("Plugin from %+v does not implement any plugin interface", plugin)
```
Comment (emilevauge) [inline plugin/manager.go]: done

**Comment by emilevauge** [comment]
@timoreimann thanks for your review :)

>Should we expand the docs?

Of course! That's why I added `[WIP]` in the description, just wanted to validate the implementation first :)

>We have no single test. Should we add some?

I will add an integration test.

>I wonder if this will affect building Traefik locally on Mac machines (which I use) where plugins are not supported yet. Will need to test.

Hum. I hope this will not be an issue... Looking forward to get the results of your tests.

**Comment by emilevauge** [comment]
OK, bad news... We can't load an external plugin if Traefik is built with `CGO_ENABLED=0`, and we **really** need this to build a statically linked golang executable to run in a Docker container.
https://github.com/golang/go/issues/19569

**Comment by emilevauge** [comment]
To be continued #1865

**Pull request closed without merging**